src/Security/OpenIdConfigurationProvider.php

```diff
@@ -72,81 +229,148 @@ public function getResourceOwnerDetailsUrl(AccessToken $token): string
 
+                foreach ($jwks['keys'] as $key) {
+                    $kid = $key['kid'];
+                    if ($key['kty'] === 'RSA') {
+                        $e = self::base64urlDecode($key['e']);
+                        $n = self::base64urlDecode($key['n']);
+                        $keys[$kid] = XMLSecurityKey::convertRSA($n, $e);
+                    } else {
+                        throw new ItkOpenIdConnectException('Unsupported key data for key id: ' . $kid);
+                    }
+                }
 
+                $item->set($keys);
+                $item->expiresAfter($this->cacheDuration);
+            }
+        } catch (InvalidArgumentException $e) {
+            // @TODO 1
+        }
+
+        return $keys;
+    }
+
+    private static function base64urlDecode($input)
+    {
+        return base64_decode(strtr($input, '-_', '+/'));
+    }
 
     /**
-     * Get Configuration.
+     * Fetch remote json resource
      *
-     * @throws \Exception
+     * @return array
+     *   Json decoded to array
+     *
+     * @throws ItkOpenIdConnectException
      */
-    private function getConfiguration(string $key): string
+    private function fetchJsonResource(string $resourceUrl): array
     {
-        $this->loadCache();
+        try {
+            $response = $this->getHttpClient()->request('GET', $resourceUrl);
+
+            if (200 !== $response->getStatusCode()) {
+                throw new ItkOpenIdConnectException('Cannot access json resource: ' .  $resourceUrl);
+            }
+
+            $content = $response->getBody()->getContents();
 
-        return $this->cache[$key];
+            return json_decode($content, true, 512, JSON_THROW_ON_ERROR);
+        } catch (GuzzleException $e) {
+            // @TODO 1
+        } catch (RuntimeException $e) {
+            // @TODO 2
+        } catch (JsonException $e) {
+            // @TODO 3
+        }
     }
 
     /**
-     * Load Cache.
+     * Get Configuration option for key.
+     *
+     * @param string $key
+     *  The configuration key
      *
-     * @throws \Exception
+     * @return string
+     *   The configuration value for the given key
      */
-    private function loadCache(): void
+    private function getConfiguration(string $key): string
     {
-        if (null === $this->cache) {
-            // refresh cache if needed
-            if (!file_exists($this->cachePath) || (time() - filemtime($this->cachePath)) > $this->cacheDuration) {
-                $this->refreshCache();
+        $cacheKey = self::CACHE_KEY . 'configuration';
+        try {
+            $item = $this->cacheItemPool->getItem($cacheKey);
+            if ($item->isHit()) {
+                $configuration = $item->get();
+            } else {
+                $configuration = $this->fetchJsonResource($this->openIDConnectMetadataUrl);
+                $item->set($configuration);
+                $item->expiresAfter($this->cacheDuration);
```

Comment:
Do you need to use the save-method on the CacheItemPool in order to persist the item?

Reply:
Documentation is unclear, but have found other examples calling `save()`. Added `$this->cacheItemPool->save($item);`